### PR TITLE
Added support for race/size/element/min level/max level quest objectives

### PR DESCRIPTION
**Pull request by Atemo**
* **Addressed Issue(s)**: https://github.com/rathena/rathena/issues/5162

* **Server Mode**: both

* **Description of Pull Request**: 

The PR adds support for race/size/element/min level/max level quest objectives

Thanks to @aleos89

### CODE REVIEW

File: db/import-tmpl/quest_db.yml [hunk 1]
```diff
@@ -30,8 +30,18 @@
 #                   Specify without "+" for the exact time the quest expires using "d" (optionnal), [0-23]"h" (required), [0-59]"mn" (optionnal), [0-59]"s" (optionnal) format.
 #                   Please note the number before "d" only shift the exact timer to the given day(s).
 #   Targets:        Quest objective target. (Default: null)
-#     - Mob         Monster to kill.
-#       Count       Amount of monsters to kill.
+#                   The target can be a couple of node Mob/Count or of Id/Race/Size/Element/MinLevel/MaxLevel.
+#                   If Mob is supplied, Count is required and the others field are ignored.
```
Comment (aleos89) [inline db/import-tmpl/quest_db.yml:34]:
```suggestion
#                   If Mob is supplied, Count is required and the other fields are ignored.
```

File: db/import-tmpl/quest_db.yml [hunk 2]
```diff
@@ -30,8 +30,18 @@
...
+#                   If Id is supplied, at least one parameter Race/Size/Element/MinLevel/MaxLevel is required.
```
Comment (aleos89) [inline db/import-tmpl/quest_db.yml:35]:
```suggestion
#                   If Id is supplied, at least one other field of Race/Size/Element/MinLevel/MaxLevel is required.
```

File: src/map/clif.cpp [hunk 3]
```diff
@@ -17168,10 +17216,14 @@ void clif_quest_send_mission(struct map_session_data *sd)
 		WFIFOW(fd, i*104+20) = static_cast<uint16>(qi->objectives.size());
 
 		for (int j = 0 ; j < qi->objectives.size(); j++) {
-			WFIFOL(fd, i*104+22+j*30) = qi->objectives[j]->mob_id;
+			struct mob_db *mob = mob_db(qi->objectives[j]->mob_id);
+
+			WFIFOL(fd, i*104+22+j*30) = (mob ? qi->objectives[j]->mob_id : 1002);
```
Comment (aleos89) [inline src/map/clif.cpp:17221]:
```suggestion
			WFIFOL(fd, i*104+22+j*30) = (mob ? qi->objectives[j]->mob_id : MOBID_PORING);
```

File: src/map/clif.cpp [hunk 4]
```diff
@@ -17099,26 +17142,30 @@ void clif_quest_send_list(struct map_session_data *sd)
...
 
 #if PACKETVER >= 20150513
 				WFIFOL(fd, offset) = sd->quest_log[i].quest_id * 1000 + j;
 				offset += 4;
-				WFIFOL(fd, offset) = 0; // TODO: Find info - mobType
+				WFIFOL(fd, offset) = (race ? race : (size ? size : (element ? element : 0)));
 				offset += 4;
 #endif
-				WFIFOL(fd, offset) = qi->objectives[j]->mob_id;
+				WFIFOL(fd, offset) = ((mob && qi->objectives[j]->mob_id > 0) ? qi->objectives[j]->mob_id : 1002);
```
Comment (aleos89) [inline src/map/clif.cpp:17153]:
```suggestion
				WFIFOL(fd, offset) = ((mob && qi->objectives[j]->mob_id > 0) ? qi->objectives[j]->mob_id : MONID_PORING);
```

File: src/map/clif.cpp [hunk 5]
```diff
@@ -17202,25 +17259,29 @@ void clif_quest_add(struct map_session_data *sd, struct quest *qd)
...
+
 #if PACKETVER >= 20150513
 		WFIFOL(fd, offset) = qd->quest_id * 1000 + i;
 		offset += 4;
-		WFIFOL(fd, offset) = 0; // TODO: Find info - mobType
+		WFIFOL(fd, offset) = (race ? race : (size ? size : (element ? element : 0)));	// effect ?
 		offset += 4;
 #endif
-		WFIFOL(fd, offset) = qi->objectives[i]->mob_id;
+		WFIFOL(fd, offset) = ((mob && qi->objectives[i]->mob_id > 0) ? qi->objectives[i]->mob_id : 1002);	// 0 can't be used as it displays "Novice" job regardless of the clif_mobtype_name
```
Comment (aleos89) [inline src/map/clif.cpp:17271]:
```suggestion
		WFIFOL(fd, offset) = ((mob && qi->objectives[i]->mob_id > 0) ? qi->objectives[i]->mob_id : MOBID_PORING);	// 0 can't be used as it displays "Novice" job regardless of the clif_mobtype_name
```

File: src/map/clif.cpp [hunk 6]
```diff
@@ -17150,7 +17150,7 @@ void clif_quest_send_list(struct map_session_data *sd)
 				WFIFOL(fd, offset) = (race ? race : (size ? size : (element ? element : 0)));
 				offset += 4;
 #endif
-				WFIFOL(fd, offset) = ((mob && qi->objectives[j]->mob_id > 0) ? qi->objectives[j]->mob_id : 1002);
+				WFIFOL(fd, offset) = ((mob && qi->objectives[j]->mob_id > 0) ? qi->objectives[j]->mob_id : MONID_PORING);
```
Comment (Balferian) [inline src/map/clif.cpp:17153]: MOBID_PORING

File: db/re/quest_db.yml [hunk 7]
```diff
@@ -2245,24 +2255,74 @@ Body:
...
+    Targets:
+      - Id: 1
+        Count: 10
+        Race: Brute
+        MinLevel: 140
   - Id: 5406
     Title: "[Repeat]Warrior Discipline-Insect"
+    Targets:
+      - Id: 1
+        Count: 10
```
Comment (attackjom) [inline db/re/quest_db.yml:2274]: I think it should be  --> Count: 50

**Comment by Atemo** [comment]
@attackjom 
For our `[Repeat]Warrior Discipline-Animal` quest on 2020 client I set arbitrarily the max level to 175
screenrAthena006
On iRO they display `or highter` instead of `175`
screenChaos006

Do you know how the `MaxLevel` is displayed on kRO?

**Comment by attackjom** [comment]
> @attackjom
> For our `[Repeat]Warrior Discipline-Animal` quest on 2020 client I set arbitrarily the max level to 175
> screenrAthena006
> On iRO they display `or highter` instead of `175`
> screenChaos006
> 
> Do you know how the `MaxLevel` is displayed on kRO?

I'll test it now. please wait.

**Comment by attackjom** [comment]
@Atemo 

Brute race
image

**Comment by Balferian** [comment]
`#define MAX_LEVEL 200`
@Atemo some monsters have higher lvl then 200.
Maybe add new variable special for quests?
for example: https://www.divine-pride.net/database/monster/20616

**Comment by Atemo** [comment]
The target maxlevel had been changed in the last commit, it doesn't take max_level of the server now, it's 0 by default (objective true) when minlevel is defined and maxlevel undef.

**Pull request merged**